crates/ruff_linter/src/suppression.rs

```diff
@@ -0,0 +1,905 @@
+
+    /// Range containing the reason for the suppression
+    reason: TextRange,
+}
+
+impl SuppressionComment {
+    /// Return the suppressed codes as strings
+    fn codes_as_str(&self, source: &str) -> Vec<String> {
+        self.codes
+            .iter()
+            .map(|range| source.slice(range).to_string())
+            .collect()
+    }
+
+    /// Whether the comment "matches" another comment, based on indentation and suppressed codes
+    fn matches(&self, other: &SuppressionComment, source: &str) -> bool {
+        ((self.action == SuppressionAction::Enable && other.action == SuppressionAction::Disable)
+            || (self.action == SuppressionAction::Disable
+                && other.action == SuppressionAction::Enable))
+            && self.indent == other.indent
+            && self.codes_as_str(source) == other.codes_as_str(source)
+    }
+}
+
+#[allow(unused)]
+#[derive(Clone, Debug)]
+pub(crate) struct Suppression {
+    /// The lint code being suppressed
+    code: String,
+
+    /// Range for which the suppression applies
+    range: TextRange,
+
+    /// Any comments associated with the suppression
+    comments: SmallVec<[SuppressionComment; 2]>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum InvalidSuppressionKind {
+    /// Trailing suppression not supported
+    Trailing,
+
+    /// No matching enable or disable suppression found
+    Unmatched,
+
+    /// Suppression does not match surrounding indentation
+    Indentation,
+}
+
+#[allow(unused)]
+#[derive(Clone, Debug)]
+pub(crate) struct InvalidSuppression {
+    kind: InvalidSuppressionKind,
+    comment: SuppressionComment,
+}
+
+#[allow(unused)]
+#[derive(Debug)]
+pub(crate) struct Suppressions {
+    /// Valid suppression ranges with associated comments
+    valid: Vec<Suppression>,
+
+    /// Invalid suppression comments
+    invalid: Vec<InvalidSuppression>,
+
+    /// Parse errors from suppression comments
+    errors: Vec<ParseError>,
+}
+
+impl Suppressions {
+    pub(crate) fn from_tokens(source: &str, tokens: &Tokens) -> Suppressions {
+        let mut builder = SuppressionsBuilder::new(source);
+        builder.load_from_tokens(tokens)
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct SuppressionsBuilder<'a> {
+    source: &'a str,
+
+    valid: Vec<Suppression>,
+    invalid: Vec<InvalidSuppression>,
+    errors: Vec<ParseError>,
+
+    pending: Vec<SuppressionComment>,
+}
+
+impl<'a> SuppressionsBuilder<'a> {
+    pub(crate) fn new(source: &'a str) -> Self {
+        Self {
+            source,
+            ..Default::default()
+        }
+    }
+
+    fn match_comments(&mut self, current_indent: &String) {
+        let mut comment_index = 0;
+
+        // for each pending comment, search for matching comments at the same indentation level,
+        // generate range suppressions for any matches, and then discard and unmatched comments
+        // from the outgoing indentation block
+        while comment_index < self.pending.len() {
```

Comment:
I wonder if `load_from_tokens` should store a `Vec<(&str, Vec<SuppressionComment>)` where the first item is the indent at a given level, and the second item is the suppression comments at that level.

I think this method would greatly simplify because you could just iterate over the comments at the given level instead of having to match them by indent again. 

The downside is that we allocate a few extra `Vec`'s, one for each level with a suppression comment. I think that's fine. 


Making this change should also remove the need for `PendingSuppressionComment` because we no longer need to store the `indent` anywhere (I think)